File: app/src/main/java/com/micoyc/speakthat/AppListAdapter.java
package com.micoyc.speakthat;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageButton;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AppListAdapter extends RecyclerView.Adapter<AppListAdapter.ViewHolder> {
    private List<FilterSettingsActivity.AppFilterItem> items;
    private FilterSettingsActivity.OnAppActionListener removeListener;
    private FilterSettingsActivity.OnAppActionListener privateToggleListener;
    private FilterSettingsActivity.OnAppActionListener editListener;
    private final boolean showPrivateToggle;
    private final Map<String, String> appNameCache = new HashMap<>();

    public AppListAdapter(List<FilterSettingsActivity.AppFilterItem> items, 
                         FilterSettingsActivity.OnAppActionListener removeListener,
                         FilterSettingsActivity.OnAppActionListener privateToggleListener,
                         FilterSettingsActivity.OnAppActionListener editListener,
                         boolean showPrivateToggle) {
        this.items = items;
        this.removeListener = removeListener;
        this.privateToggleListener = privateToggleListener;
        this.editListener = editListener;
        this.showPrivateToggle = showPrivateToggle;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_app_filter, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        FilterSettingsActivity.AppFilterItem item = items.get(position);
        
        holder.textAppName.setText(resolveDisplayName(holder.itemView.getContext(), item.packageName));
        // Show/hide private toggle based on caller intent
        holder.checkBoxPrivate.setVisibility(showPrivateToggle ? View.VISIBLE : View.GONE);
        if (showPrivateToggle) {
            holder.checkBoxPrivate.setChecked(item.isPrivate);
        } else {
            // Avoid stale listeners when reused in recycled views
            holder.checkBoxPrivate.setOnCheckedChangeListener(null);
        }
        
        // Make text clickable for editing
        holder.textAppName.setOnClickListener(v -> {
            if (editListener != null) {
                int pos = holder.getBindingAdapterPosition();
                if (pos != RecyclerView.NO_POSITION) {
                    editListener.onAction(pos);
                }
            }
        });
        
        // Set up checkbox listener
        if (showPrivateToggle) {
            holder.checkBoxPrivate.setOnCheckedChangeListener(null); // Clear previous listener
            holder.checkBoxPrivate.setOnCheckedChangeListener((buttonView, isChecked) -> {
                if (privateToggleListener != null) {
                    int pos = holder.getBindingAdapterPosition();
                    if (pos != RecyclerView.NO_POSITION) {
                        privateToggleListener.onAction(pos);
                    }
                }
            });
        }
        
        // Set up remove button listener
        holder.buttonRemove.setOnClickListener(v -> {
            if (removeListener != null) {
                int pos = holder.getBindingAdapterPosition();
                if (pos != RecyclerView.NO_POSITION) {
                    removeListener.onAction(pos);
                }
            }
        });
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        TextView textAppName;
        CheckBox checkBoxPrivate;
        ImageButton buttonRemove;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            textAppName = itemView.findViewById(R.id.textAppName);
            checkBoxPrivate = itemView.findViewById(R.id.checkBoxPrivate);
            buttonRemove = itemView.findViewById(R.id.buttonRemove);
        }
    }

    private String resolveDisplayName(android.content.Context context, String packageName) {
        if (appNameCache.containsKey(packageName)) {
            return appNameCache.get(packageName);
        }

        String displayName = packageName;
        PackageManager pm = context.getPackageManager();
        try {
            ApplicationInfo appInfo = pm.getApplicationInfo(packageName, 0);
            CharSequence label = pm.getApplicationLabel(appInfo);
            if (label != null && label.length() > 0) {
                displayName = label.toString();
            }
        } catch (PackageManager.NameNotFoundException ignored) {
            AppListData appData = AppListManager.INSTANCE.findAppByPackage(context, packageName);
            if (appData != null && appData.displayName != null && !appData.displayName.isEmpty()) {
                displayName = appData.displayName;
            }
        }

        appNameCache.put(packageName, displayName);
        return displayName;
    }
} 

File: app/src/main/java/com/micoyc/speakthat/AppListAdapter.kt
package com.micoyc.speakthat

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Filter
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.NonNull
import androidx.annotation.Nullable

/**
 * Adapter for displaying app list items from JSON data in AutoCompleteTextView
 */
class AppSearchAdapter(
    context: Context,
    private val appList: List<AppListData>
) : ArrayAdapter<AppListData>(context, R.layout.item_app_selector, appList) {
    
    private val inflater = LayoutInflater.from(context)
    private var filteredApps = ArrayList(appList.take(20)) // Start with first 20 apps
    
    override fun getCount(): Int = filteredApps.size
    
    override fun getItem(position: Int): AppListData? = 
        if (position < filteredApps.size) filteredApps[position] else null
    
    @NonNull
    override fun getView(position: Int, @Nullable convertView: View?, @NonNull parent: ViewGroup): View {
        val view = convertView ?: inflater.inflate(R.layout.item_app_selector, parent, false)
        
        val app = getItem(position)
        if (app != null) {
            // Set app icon (using default icon since we don't have actual app icons)
            val iconView = view.findViewById<ImageView>(R.id.appIcon)
            iconView.setImageResource(R.drawable.ic_app_default)
            
            // Set app name
            val nameView = view.findViewById<TextView>(R.id.appName)
            nameView.text = app.displayName
            
            // Set package name
            val packageView = view.findViewById<TextView>(R.id.packageName)
            packageView.text = app.packageName
            
            // Set category as subtitle (if you want to show it)
            // You can add a category TextView to the layout if needed
        }
        
        return view
    }
    
    override fun getFilter(): Filter = AppFilter()
    
    private inner class AppFilter : Filter() {
        override fun performFiltering(constraint: CharSequence?): FilterResults {
            val results = FilterResults()
            val suggestions = ArrayList<AppListData>()
            
            val filterText = constraint?.toString()?.lowercase()?.trim() ?: ""
            
            if (filterText.isEmpty()) {
                // Show first 20 apps when no filter
                suggestions.addAll(appList.take(20))
            } else {
                // Filter apps based on query
                for (app in appList) {
                    if (app.matchesQuery(filterText)) {
                        suggestions.add(app)
                    }
                }
            }
            
            results.values = suggestions
            results.count = suggestions.size
            
            // Debug logging
            android.util.Log.d("AppSearchAdapter", "Filtering with '$filterText' - found ${suggestions.size} results")
            
            return results
        }
        
        override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
            filteredApps.clear()
            if (results?.values != null) {
                @Suppress("UNCHECKED_CAST")
                filteredApps.addAll(results.values as List<AppListData>)
            }
            
            // Debug logging
            android.util.Log.d("AppSearchAdapter", "Publishing ${filteredApps.size} results")
            
            // Just notify the adapter of changes
            notifyDataSetChanged()
        }
    }
} 